KAFKA-16077: Streams with state updater fails to close task upon fencing

lucasbru opened the pull request:
Streams fails to close task when the producer is fenced, but input partitions are updated in a new assignment happening at the same time.

There is a race condition in the state updater that can cause the following:

 1. We have an active task in the state updater
 2. We get fenced. We recreate the producer, transactions now uninitialized. We ask the state updater to give back the task, add a pending action to close the task clean once it’s handed back
 3. We get a new assignment with updated input partitions. The task is still owned by the state updater, so we ask the state updater again to hand it back and add a pending action to update its input partition
 4. The task is handed back by the state updater. We update its input partitions but forget to close it clean (pending action was overwritten)
 5. Now the task is in an initialized state, but the underlying producer does not have transactions initialized

This can cause an IllegalStateException: `Invalid transition attempted from state UNINITIALIZED to state IN_TRANSACTION` when running in EOSv2.

To fix this, we introduce a new pending action CloseReviveAndUpdateInputPartitions that is added when we handle a new assignment with updated input partitions, but we still need to close the task before reopening it.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

lucasbru committed:
KAFKA-16077: Streams fails to close task after restoration when input partitions are updated

There is a race condition in the state updater that can cause the following:

 1. We have an active task in the state updater
 2. We get fenced. We recreate the producer, transactions now uninitialized. We ask the state updater to give back the task, add a pending action to close the task clean once it’s handed back
 3. We get a new assignment with updated input partitions. The task is still owned by the state updater, so we ask the state updater again to hand it back and add a pending action to update its input partition
 4. The task is handed back by the state updater. We update its input partitions but forget to close it clean (pending action was overwritten)
 5. Now the task is in an initialized state, but the underlying producer does not have transactions initialized

This can cause an IllegalStateException: `Invalid transition attempted from state UNINITIALIZED to state IN_TRANSACTION` when running in EOSv2.

To fix this, we introduce a new pending action CloseReviveAndUpdateInputPartitions that is added when we handle a new assignment with updated input partitions, but we still need to close the task before reopening it.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1164,9 +1243,12 @@ public void shouldHandleMultipleRemovedTasksFromStateUpdater() {
             .inState(State.RESTORING)
             .withInputPartitions(taskId03Partitions).build();
+        final StreamTask taskToCloseReviveAndUpdateInputPartitions = statefulTask(taskId04, taskId04ChangelogPartitions)
+                .inState(State.RESTORING)
+                .withInputPartitions(taskId04Partitions).build();
```
nit:
```suggestion
            .inState(State.RESTORING)
            .withInputPartitions(taskId04Partitions).build();
```

lucasbru replied:
Done

cadonna replied:
This does not seem done. But nevermind, if it is just this I will of course not block the PR.

cadonna replied:
I think there was a similar indentation issue on line 1713 that you fixed. I also saw one on line 1652.

lucasbru replied:
Right, I only fixed the other one. I think I recently changed my IDE to use 8 character continuation indentation, became somebody complained about it in a client PR. Changing it back. Honestly, I wish we'd just have automatic formatting in kafka.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -950,6 +963,12 @@ private void handleRestoredTasksFromStateUpdater(final long now,
+            } else if ((inputPartitions = tasks.removePendingTaskToCloseReviveAndUpdateInputPartitions(task.id())) != null) {
+                if (closeTaskClean(task, tasksToCloseDirty, taskExceptions)) {
+                    task.revive();
+                    task.updateInputPartitions(inputPartitions, topologyMetadata.nodeToSourceTopics(task.id()));
```
My comment above also applies here.

lucasbru replied:
Done

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -916,6 +923,12 @@ private void handleRemovedTasksFromStateUpdater() {
+                if (closeTaskClean(task, tasksToCloseDirty, taskExceptions)) {
+                    task.revive();
+                    task.updateInputPartitions(inputPartitions, topologyMetadata.nodeToSourceTopics(task.id()));
+                    stateUpdater.add(task);
```
Where is the task initialized?
The task is closed cleanly and the call to `revive()` sets the task to `CREATED`. However, the task is never initialized after that.

lucasbru replied:
Good point! I was supposed to call `addTaskToStateUpdater` here instead.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1127,6 +1175,36 @@ public void shouldUpdateInputPartitionsOfTasksRemovedFromStateUpdater() {
 
+    @Test
+    public void shouldCloseReviveAndUpdateInputPartitionsOfTasksRemovedFromStateUpdater() {
+        final StreamTask task00 = statefulTask(taskId00, taskId00ChangelogPartitions)
```
nit:
Could you please use more meaningful names for those task variables? Something like `activeTask` and `standbyTask` would already be better.

lucasbru replied:
Done

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1127,6 +1175,36 @@ public void shouldUpdateInputPartitionsOfTasksRemovedFromStateUpdater() {
+        final TasksRegistry tasks = mock(TasksRegistry.class);
+        when(tasks.removePendingTaskToRecycle(any())).thenReturn(null);
+        when(tasks.removePendingTaskToCloseReviveAndUpdateInputPartitions(task00.id())).thenReturn(taskId02Partitions);
+        when(tasks.removePendingTaskToCloseReviveAndUpdateInputPartitions(task01.id())).thenReturn(taskId03Partitions);
```
I thought a standby task can never be added to this pending action.

lucasbru replied:
Done

cadonna reviewed (commented):
Thanks for the fix, @lucasbru !

I have one major question about initialization of tasks.

Should we also add a integration test that tests the scenario?

lucasbru committed:
address comments

lucasbru reviewed (commented):
Thanks for the review @cadonna . I addressed all comments. Good catch about the initialization

cadonna reviewed (approved):
Thanks @lucasbru !

LGTM!

There are just indentation issues, but if you do not need to restart the builds, let's leave those for a follow-up PR.

cadonna commented:
@lucasbru What do you think about my previous question?

> Should we also add a integration test that tests the scenario?

lucasbru commented:
> @lucasbru What do you think about my previous question?
> 
> > Should we also add a integration test that tests the scenario?

We don't test the other cases of the pending actions, but I agree that it would be good to test this. I thought about it for a bit and I think it will require quite some tooling to get it tested in an integration test (maybe I'm missing something though).

Since we have covered it by unit tests, and this is a blocker issue for 3.7, I'd vote for merging the fix and creating a follow-up ticket for the integration test.

lucasbru committed:
formatting

lucasbru committed:
fix

We should not remove the task twice, otherwise, we'll end up in this situation

1. We have an active task in the state updater
2. We get fenced. We recreate the producer, transactions now uninitialized. We ask the state updater to give back the task, add a pending action to close the task clean once it’s handed back
3. The state updater moves the task from the updating tasks to the removed tasks
4. We get a new assignment with updated input partitions. The task is still owned by the state updater, so we ask the state updater again to hand it back (adding a task+remove into the task and action queue) and add a pending action to close, revive and update input partitions
5. The task is handed back by the state updater. We close revive and update input partitions, and add the task back to the state updater
6. The state updater executes the "task+remove" action that is still in its task + action queue, and hands the task immediately back to the main thread
7. The main thread discoveres a removed task that was not restored and has no pending action attached to it. IllegalStateException

lucasbru committed:
fix test
